### ISSUE TITLE
salt and salt-call do not return the same shell grain

### ISSUE THREAD

**Issue by ealphonse** [saltstack/salt#32989]
### Description of Issue/Question

The shell grain used to define which shell binary to call when using cmd.shell (cmd.run) differs depending on the way one calls the command. This breaks some states written for bash and not for sh.

We work around it by using `cmd.script` instead but it would be nice to fix this discrepancy as it is hard to spot.

Is there a minion option overriding the default value?
### Setup

We have ubuntu 14.04 LTS and `/bin/bash` as the default shell everywhere (root included).
### Steps to Reproduce Issue

From a shell interpreter:

```
$ hostname
sarje-infra
$ salt-call grains.get shell
local:
    /bin/bash
# and from the master
$ salt sarje-infra grains.get shell
sarje-infra:
    /bin/sh
```

We have the same result on all minions.
### Versions Report

Same version for master and minion:

```
Salt Version:
           Salt: 2015.8.8

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: Not Installed
           Mako: 0.9.1
         PyYAML: 3.10
          PyZMQ: 14.0.1
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.6
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-24-generic
         system: Ubuntu 14.04 trusty
```

**Comment by hexedpackets** [comment]
I was bitten by this as well. For me it came up as a difference between using the cmd.run state (/bin/sh) and the cmd.run module (/bin/bash).

**Comment by jfindlay** [comment]
@ealphonse, @hexedpackets, thanks for reporting.  The `shell` grain is derived from the `$SHELL` environment variable ([`return {'shell': os.environ.get('SHELL', '/bin/sh')}`](https://github.com/saltstack/salt/blob/v2016.3.0rc2/salt/grains/extra.py#L18-L24)).  I am not sure what the best thing to do here is because remote management through the minion daemon (`salt '*' grains.item shell`) may use a different environment than `salt-call` which will inherit the environment of it's enclosing shell, which is not really deterministic in general at least because of the variety of contexts an arbitrary shell command can be executed in.

**Comment by ealphonse** [comment]
Hi @jfindlay, thank you for investigating the issue.

Indeed, the env is minimal with remote management:
```
# salt sarje-infra cmd.run 'env'
sarje-infra:
    UPSTART_INSTANCE=
    UPSTART_JOB=salt-minion
    TERM=linux
    PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
    LC_ALL=C
    PWD=/root
```

It is os-specific, but if I understand correctly, `start-stop-daemon --background` used to launch the service on debian-based distros does not pass the env to the process. One has to use e.g. [daemon](http://libslack.org/daemon/manpages/daemon.1.html) for that, and then we can make use of `/etc/default/salt-minion` to specify the correct env, but again probably too os-specific.

What about a minion config option to specify the default shell instead of hard-coding `/bin/sh`?

**Comment by jfindlay** [comment]
@ealphonse, that seems like it could be a good solution.  See also #32806.

**Comment by cachedout** [comment]
Really interesting discussion here.

I tend to generally be of the opinion that Salt grains should accurately reflect the environment that they're running under, even if this means that there's some discrepancy between their output when called from one tool versus another. Trying to massage this feels like a little bit of a black hole to me.

I'm more curious about real-world cases where this is happening. Are there existing salt module/state pairing where this is especially prevalent? Or is this just an idiosyncrasy that's biting people in custom states? If we can attack the problem at that layer, it feels like a safer approach to me.

**Comment by ealphonse** [comment]
Hello @cachedout,

> Salt grains should accurately reflect the environment that they're running under,

I totally agree, and this is expected from the documentation. However, as far as I understand, it seems this environment cannot be configured:
```
# salt sarje-infra cmd.run 'env'
sarje-infra:
    UPSTART_INSTANCE=
    UPSTART_JOB=salt-minion
    TERM=linux
    PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
    LC_ALL=C
    PWD=/root
```

Note that the debian-based service does not set `SHELL` and I did not find a way to make it so. So basically it always is the hard-coded value: `/bin/sh`.

How can we make salt-minion aware of its shell?

As for the use-case, nothing special here, testing locally with salt-call and then executing remotely.

**Comment by cachedout** [comment]
Hi @ealphonse 

Yes, I think I agree with you here. Though, I do see a reference to an `EnvironmentFile` for systemd in the Debian package. https://github.com/saltstack/salt/blob/develop/pkg/deb/salt-minion.service#L6

Could setting a shell there be a way out of this?

**Comment by ealphonse** [comment]
Hello @cachedout,

I already tried configuring the shell env in `/etc/default/salt-minion`, which is also read by `/etc/init.d/salt-minion`, but it cannot work because `start-stop-daemon --background`does not pass the env, as noted above.

I do not know and use `systemd` though. Do you think this is the fix? Is the service called different from `/etc/init.d/salt-minion`?

**Comment by wryfi** [comment]
We were having this issue on our ubuntu-14.04 systems. We were able to fix it by dropping a file on our minions in /etc/default/salt-minion, with `export SHELL=/bin/bash`.

**N.B. #31202!** We created states that **remove** the legacy /etc/init.d scripts, so our systems exclusively use the upstart job (both are still in the package, for some unknown reason). This avoids the issue (mentioned above) with `start-stop-daemon` failing to pass the environment along, as the upstart job does not use `start-stop-daemon`, but rather starts the daemon directly.
